In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler, OneHotEncoder 
from path import Path
from uszipcode import Zipcode #pip install uszipcode



In [22]:
file_path = Path("../Resources/sanFranCrime.csv")
initial_df = pd.read_csv(file_path)
initial_df.head()

IncidntNum      Category                                        Descript  \
0   120058272   WEAPON LAWS                       POSS OF PROHIBITED WEAPON   
1   120058272   WEAPON LAWS  FIREARM, LOADED, IN VEHICLE, POSSESSION OR USE   
2   141059263      WARRANTS                                  WARRANT ARREST   
3   160013662  NON-CRIMINAL                                   LOST PROPERTY   
4   160002740  NON-CRIMINAL                                   LOST PROPERTY   

  DayOfWeek                    Date   Time  PdDistrict      Resolution  \
0    Friday  01/29/2016 12:00:00 AM  11:00    SOUTHERN  ARREST, BOOKED   
1    Friday  01/29/2016 12:00:00 AM  11:00    SOUTHERN  ARREST, BOOKED   
2    Monday  04/25/2016 12:00:00 AM  14:59     BAYVIEW  ARREST, BOOKED   
3   Tuesday  01/05/2016 12:00:00 AM  23:50  TENDERLOIN            NONE   
4    Friday  01/01/2016 12:00:00 AM  00:30     MISSION            NONE   

                  Address           X          Y  \
0  800 Block of BRYANT ST -122.403405  37.775421   
1  800 Block of BRYANT ST -122.403405  37.775421   
2   KEITH ST / SHAFTER AV -122.388856  37.729981   
3  JONES ST / OFARRELL ST -122.412971  37.785788   
4    16TH ST / MISSION ST -122.419672  37.765050   

                                Location            PdId  
0   (37.775420706711, -122.403404791479)  12005827212120  
1   (37.775420706711, -122.403404791479)  12005827212168  
2  (37.7299809672996, -122.388856204292)  14105926363010  
3  (37.7857883766888, -122.412970537591)  16001366271000  
4  (37.7650501214668, -122.419671780296)  16000274071000

In [23]:
#create separate dataframe for location data
geo_data_df = pd.DataFrame(data=initial_df, columns=["IncidntNum","Date","Address","X","Y","Location"])
geo_data_df.head()

IncidntNum                    Date                 Address           X  \
0   120058272  01/29/2016 12:00:00 AM  800 Block of BRYANT ST -122.403405   
1   120058272  01/29/2016 12:00:00 AM  800 Block of BRYANT ST -122.403405   
2   141059263  04/25/2016 12:00:00 AM   KEITH ST / SHAFTER AV -122.388856   
3   160013662  01/05/2016 12:00:00 AM  JONES ST / OFARRELL ST -122.412971   
4   160002740  01/01/2016 12:00:00 AM    16TH ST / MISSION ST -122.419672   

           Y                               Location  
0  37.775421   (37.775420706711, -122.403404791479)  
1  37.775421   (37.775420706711, -122.403404791479)  
2  37.729981  (37.7299809672996, -122.388856204292)  
3  37.785788  (37.7857883766888, -122.412970537591)  
4  37.765050  (37.7650501214668, -122.419671780296)

In [24]:
# Import/setup dependencies for uszipcode
# https://uszipcode.readthedocs.io/index.html#example-usage
# This took about 10 min to run FYI
from uszipcode import Zipcode, SearchEngine
search = SearchEngine(simple_zipcode=True)

def convertToZip(df):
    '''Creates a zipcode column from Y and X coordinates in a           dataframe. uszipcode returns multiple datapoints by default so      only selecting the zipcode in the return value'''
    result = search.by_coordinates(df['Y'], df['X'])
    return result[0].zipcode

geo_data_df['Zipcode'] = geo_data_df.apply(convertToZip, axis=1)


In [26]:
# Check our df to see if the Zipcodes look ok
geo_data_df.head(10)

IncidntNum                    Date                       Address  \
0    120058272  01/29/2016 12:00:00 AM        800 Block of BRYANT ST   
1    120058272  01/29/2016 12:00:00 AM        800 Block of BRYANT ST   
2    141059263  04/25/2016 12:00:00 AM         KEITH ST / SHAFTER AV   
3    160013662  01/05/2016 12:00:00 AM        JONES ST / OFARRELL ST   
4    160002740  01/01/2016 12:00:00 AM          16TH ST / MISSION ST   
5    160002869  01/01/2016 12:00:00 AM         1700 Block of BUSH ST   
6    160003130  01/02/2016 12:00:00 AM           MARY ST / HOWARD ST   
7    160003259  01/02/2016 12:00:00 AM          200 Block of EDDY ST   
8    160003970  01/02/2016 12:00:00 AM             4TH ST / BERRY ST   
9    160003641  01/01/2016 12:00:00 AM       100 Block of CAMERON WY   
10   160086863  01/29/2016 12:00:00 AM         1200 Block of 19TH AV   
11   160004053  01/02/2016 12:00:00 AM         2200 Block of 21ST AV   
12   160073014  01/25/2016 12:00:00 AM  200 Block of HUNTERSPOINT BL   
13   140776777  09/15/2016 12:00:00 AM       SILVER AV / CAMELLIA AV   
14   160004069  01/02/2016 12:00:00 AM       100 Block of SANSOME ST   
15   160004150  01/02/2016 12:00:00 AM        900 Block of MARKET ST   
16   160004241  01/02/2016 12:00:00 AM       900 Block of POTRERO AV   
17   160004558  01/02/2016 12:00:00 AM         VALENCIA ST / 26TH ST   
18   160004655  01/02/2016 12:00:00 AM        900 Block of GENEVA AV   
19   160004837  01/02/2016 12:00:00 AM        900 Block of MARKET ST   

             X          Y                               Location Zipcode  
0  -122.403405  37.775421   (37.775420706711, -122.403404791479)   94103  
1  -122.403405  37.775421   (37.775420706711, -122.403404791479)   94103  
2  -122.388856  37.729981  (37.7299809672996, -122.388856204292)   94124  
3  -122.412971  37.785788  (37.7857883766888, -122.412970537591)   94108  
4  -122.419672  37.765050  (37.7650501214668, -122.419671780296)   94102  
5  -122.426077  37.788019   (37.788018555829, -122.426077177375)   94109  
6  -122.405721  37.780879  (37.7808789360214, -122.405721454567)   94103  
7  -122.411778  37.783981  (37.7839805592634, -122.411778295992)   94103  
8  -122.393357  37.775788  (37.7757876218293, -122.393357241451)   94158  
9  -122.387182  37.720967  (37.7209669615499, -122.387181635995)   94124  
10 -122.477377  37.764478  (37.7644781578695, -122.477376524003)   94122  
11 -122.477960  37.745739  (37.7457389429655, -122.477960327299)   94116  
12 -122.376758  37.735697   (37.7356970275482, -122.37675765553)   94124  
13 -122.432326  37.729271  (37.7292705199592, -122.432325871028)   94112  
14 -122.400909  37.791643    (37.791642982384, -122.40090869889)   94104  
15 -122.408595  37.783707  (37.7837069301545, -122.408595110869)   94103  
16 -122.406870  37.757290  (37.7572895904578, -122.406870402082)   94110  
17 -122.420355  37.748906  (37.7489063051829, -122.420354780861)   94110  
18 -122.439910  37.715765   (37.715765426995, -122.439909766772)   94112  
19 -122.408421  37.783570  (37.7835699386918, -122.408421116922)   94103

In [4]:
#save geo_data_df as csv
geo_data_df.to_csv(r'/Users/bkirton/Desktop/sanFranCrimePredictor/Resources/geo_data.csv')

In [5]:
#drop location data, date, dayofweek from initial_df
initial_df = initial_df.drop(columns=["Address","X","Y","Location","Date","DayOfWeek"])
initial_df.head()

IncidntNum      Category                                        Descript  \
0   120058272   WEAPON LAWS                       POSS OF PROHIBITED WEAPON   
1   120058272   WEAPON LAWS  FIREARM, LOADED, IN VEHICLE, POSSESSION OR USE   
2   141059263      WARRANTS                                  WARRANT ARREST   
3   160013662  NON-CRIMINAL                                   LOST PROPERTY   
4   160002740  NON-CRIMINAL                                   LOST PROPERTY   

    Time  PdDistrict      Resolution            PdId  
0  11:00    SOUTHERN  ARREST, BOOKED  12005827212120  
1  11:00    SOUTHERN  ARREST, BOOKED  12005827212168  
2  14:59     BAYVIEW  ARREST, BOOKED  14105926363010  
3  23:50  TENDERLOIN            NONE  16001366271000  
4  00:30     MISSION            NONE  16000274071000

In [6]:
#remaining columns will be our features. Target column is resolution
initial_df.nunique()

IncidntNum    116699
Category          39
Descript         726
Time            1439
PdDistrict        10
Resolution        14
PdId          150500
dtype: int64

In [7]:
#get number of values for resolution column
initial_df["Resolution"].value_counts()

NONE                                      107780
ARREST, BOOKED                             39416
UNFOUNDED                                   1608
JUVENILE BOOKED                             1056
EXCEPTIONAL CLEARANCE                        371
ARREST, CITED                                144
CLEARED-CONTACT JUVENILE FOR MORE INFO        58
NOT PROSECUTED                                22
LOCATED                                       20
PSYCHOPATHIC CASE                             17
JUVENILE CITED                                 3
COMPLAINANT REFUSES TO PROSECUTE               2
JUVENILE DIVERTED                              2
PROSECUTED BY OUTSIDE AGENCY                   1
Name: Resolution, dtype: int64

In [8]:
initial_df["Descript"].value_counts()

GRAND THEFT FROM LOCKED AUTO                      17741
LOST PROPERTY                                      4596
AIDED CASE, MENTAL DISTURBED                       4566
PETTY THEFT OF PROPERTY                            4416
MALICIOUS MISCHIEF, VANDALISM                      4262
                                                  ...  
POLICE BROADCAST, INTERCEPTION TO COMMIT CRIME        1
PETTY THEFT MOTORCYCLE STRIP                          1
MONEY LAUNDERING                                      1
TRANSPORTATION OF METHADONE                           1
FORCIBLE RAPE, ARMED WITH A GUN                       1
Name: Descript, Length: 726, dtype: int64

In [9]:
#bucket everything besides NONE and arrest,booked for now to create binary classification

#bucket resolutions
resolution_counts = initial_df["Resolution"].value_counts()

replace_resolution = list(resolution_counts[resolution_counts<1700].index)

for resolution in replace_resolution:
    initial_df["Resolution"] = initial_df["Resolution"].replace(resolution, "other")
    
initial_df["Resolution"].value_counts()

NONE              107780
ARREST, BOOKED     39416
other               3304
Name: Resolution, dtype: int64

Out of 150,500 rows only 3,304 are "other" after bucketing the "Resolution" column. This is only 2% of our overall data. These "other" values will be dropped in order to use the "Resolution" column as a our binary classifier for machine learning. 

In [10]:
#drop "other" values from resolution column and remove other non-essential columns
cleaned_df = initial_df[initial_df.Resolution != "other"]
cleaned_df = cleaned_df.drop(columns=["IncidntNum", "PdId","Descript","Time"])
cleaned_df = cleaned_df.dropna()
cleaned_df.head()

Category  PdDistrict      Resolution
0   WEAPON LAWS    SOUTHERN  ARREST, BOOKED
1   WEAPON LAWS    SOUTHERN  ARREST, BOOKED
2      WARRANTS     BAYVIEW  ARREST, BOOKED
3  NON-CRIMINAL  TENDERLOIN            NONE
4  NON-CRIMINAL     MISSION            NONE

In [13]:
#save cleaned_df as csv
cleaned_df.to_csv(r'/Users/bkirton/Desktop/sanFranCrimePredictor/Resources/cleaned_df.csv')